tests/tests/tier0/bluechi-is-online-agent-wait/test_bluechi_is_online_agent_wait.py

```diff
@@ -0,0 +1,153 @@
+        expected_result=expected_result,
+    )
+    future.result = result
+    future.output = output
+
+
+def exec(ctrl: BluechiControllerMachine, nodes: Dict[str, BluechiAgentMachine]):
+
+    node_foo = nodes[NODE_FOO]
+
+    # Test 1: Start agent and verify --wait option returns 0 immediately
+    node_foo.systemctl.start_unit("bluechi-agent")
+    assert node_foo.wait_for_unit_state_to_be("bluechi-agent", "active")
+    LOGGER.debug("Starting test number 1 - agent should be online.")
+
+    result_future_online = ResultFuture()
+    check_thread_online = threading.Thread(
+        target=check_agent,
+        args=(
+            node_foo.bluechi_is_online,
+            result_future_online,
+            "Checking if agent is online",
+            0,
+            0,
+        ),
+    )
+    check_thread_online.start()
+    check_thread_online.join()
+    assert result_future_online.result == 0, "Expected agent to be online"
+
+    # Test 2: Stop agent and verify --wait returns 1 after wait time is expired.
+    node_foo.systemctl.stop_unit("bluechi-agent")
+    assert node_foo.wait_for_unit_state_to_be("bluechi-agent", "inactive")
+    LOGGER.debug("Starting test number 2 - agent should remain offline.")
+
+    result_future_offline = ResultFuture()
+    check_thread_offline = threading.Thread(
+        target=check_agent,
+        args=(
+            node_foo.bluechi_is_online,
+            result_future_offline,
+            "Checking if agent remains offline",
+            5000,
+            1,
+        ),
+    )
+    check_thread_offline.start()
+    check_thread_offline.join()
+    assert result_future_offline.result != 0, "Expected agent to remain offline"
+
+    # Test 3: Stop agent, run 'bluechi-is-online agent --wait', start agent and verify --wait returns 0 before the
+    # wait time expires
+    LOGGER.debug(
+        "Starting test number 3, stopping agent to ensure it is inactive before starting `bluechi-is-online` "
+        "thread."
+    )
+    node_foo.systemctl.stop_unit("bluechi-agent")
+    assert node_foo.wait_for_unit_state_to_be("bluechi-agent", "inactive")
+    LOGGER.debug("Agent confirmed inactive after stopping.")
+
+    # Add a short delay to ensure the "inactive" state is fully processed
+    time.sleep(1)
+    LOGGER.debug("Delay to ensure agent is fully offline before starting thread.")
+
+    # Start the `bluechi-is-online` thread
+    result_future_wait = ResultFuture()
+    start_time = time.time()
+    LOGGER.debug("Starting `bluechi-is-online` thread with wait time of 20 seconds.")
+    check_thread_wait = threading.Thread(
+        target=check_agent,
+        args=(
+            node_foo.bluechi_is_online,
+            result_future_wait,
+            "Checking if agent comes online",
+            20000,
+            0,
+        ),
+    )
+    check_thread_wait.start()
+
+    # Restart the agent after a delay to simulate it coming online during the --wait period
+    time.sleep(5)
```

Comment:
5 seconds is quite a lot of time, considering we already approaching the 100 test count, and there will be more of these... It's okay to use sleep when needed, but we should really try to keep our tests as short as possible